src/ZEO/ClientStorage.py

```diff
@@ -1388,6 +1389,14 @@ def verify_cache(self, server):
         for oid, tid in self._cache.contents():
             server.verify(oid, tid)
         server.endZeoVerify()
+
+        if server_tid > self._cache.getLastTid():
+            # We verified the cache, and got no new invalidations
+            # while doing so.  The records in the cache are valid,
+            # in that invalid current records were invalidated,
+            # but the last tid is wrong.  Let's fix it:
+            self._cache.setLastTid(server_tid)
```

Comment:
``verify_cache`` called on *connection* (see ``notifyConnected``) when the server tid is > cache tid and when "quick" verification isn't possible because we've been disconnected for too long.

Invalidations are queued while this ``verify_cache`` runs, however, the queueing stops just before the check above.  The logic is torturous in many ways, in part because of ZEO4's use of threads for managing connections, and because the logic is spread over too many methods. ZEO5 will simplify this quite a bit, including getting rid of full cache verification. :)

The race here is that ``server.endZeoVerify`` very indirectly calls ``endVerify``, which calls ``finish_verification``, which stops invalidation queueing before the ``server_tid > self._cache.getLastTid()`` check is done.  Bonus: the indirect call happens in a different client thread!

I'll fix this.  The fix will be a little tricky.  Need to think about it. Good catch!